### PR TITLE
add: mplex

### LINKED ISSUE #241: Populate mplex stream muxer doc
## Done Criteria
- [x] Create a "What is mplex" section
- [x] Create a "How does mplex work" section
- [x] Create a "mplex in libp2p" section
  - Note usage of mplex and yamux and should not need to support mplex, unless dealing with legacy nodes

## Why Important
Page was created as a placeholder and requires additional material.

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- Adds content for mplex
- Closes #241

### Notes

- Section this document lives under will become populated with corresponding PRs

## Latest preview

Please view the latest Fleek preview [here](https://bafybeiff7dd37tqgbc4zdlj6pbkjfoxzunexp6vltiaoxbrjlnw6vayo7e.on.fleek.co/concepts/multiplex/mplex/).

### CODE REVIEW

File: content/concepts/multiplex/mplex.md [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
-mplex is a protocol developed for libp2p. The [spec](https://github.com/libp2p/specs/tree/master/mplex) defines a
-simple protocol for multiplexing that is widely supported across libp2p language implementations:
+## What is Mplex?
 
-{{< alert icon="💡" context="note">}}
-This section is incomplete, and many of the articles are stubs. To help fill in
-the gaps, please see the issues linked in each article to add your input and
-help us prioritize the outstanding work.
-{{< /alert >}}
+Mplex is a stream multiplexer designed for libp2p.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:9]:
```suggestion
mplex is a stream multiplexer designed for libp2p.
```

File: content/concepts/multiplex/mplex.md [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
+It is a simple protocol that does not provide many features offered by other
+stream multiplexers. Notably, mplex does not provide backpressure at the protocol
+level.
+
+Mplex runs over a reliable, ordered pipe between two peers, such as a TCP socket
+or a Unix pipe. Peers can open, write to, close, and reset a stream by sending messages
+with the appropriate [flag value](https://github.com/libp2p/specs/tree/master/mplex#flag-values).
+
+Every message (one byte) in mplex consists of a header and a length of prefixed data
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:18]: I don't understand the "one byte" part.

File: content/concepts/multiplex/mplex.md [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
 
-{{< alert icon="💡" context="note">}}
-This section is incomplete, and many of the articles are stubs. To help fill in
-the gaps, please see the issues linked in each article to add your input and
-help us prioritize the outstanding work.
-{{< /alert >}}
+Mplex is a stream multiplexer designed for libp2p.
+It is a simple protocol that does not provide many features offered by other
+stream multiplexers. Notably, mplex does not provide backpressure at the protocol
+level.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:12]: Make a separate paragraph that lists the drawbacks:
* no backpressure on data sent on streams
* no limits on how many streams the peer can open

**Review by marten-seemann — COMMENTED** [review]
Please also add a section saying that until recently, the reason mplex was still supported was compatibility with js-libp2p, which didn't have yamux support yet. Now that js-libp2p has gained yamux support (link to the release), there's really no good reason to still support mplex, unless you need to deal with legacy nodes. For all other use cases, yamux would be the better choice.

File: content/concepts/multiplex/mplex.md (hunk 3 above)
Comment (DannyS03) [reply content/concepts/multiplex/mplex.md:12]: great

File: content/concepts/multiplex/mplex.md [hunk 4]
```diff
@@ -1,14 +1,48 @@
...
+or a Unix pipe. Peers can open, write to, close, and reset a stream by sending messages
+with the appropriate [flag value](https://github.com/libp2p/specs/tree/master/mplex#flag-values).
+
+Every message in mplex consists of a header and a length of prefixed data
+segment.
+
+### Drawbacks
+
+mplex does not support backpressure because it is designed to be more flexible and
+extensible than other multiplexing protocols. Instead of using a flow control mechanism
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:24]: It's not really flexible and extensible. I think it was primarily designed for implementation simplicity, without really considering that backpressure is kind of essential when writing a networking stack. Of course we shouldn't put it like this in our docs ;)

File: content/concepts/multiplex/mplex.md [hunk 5]
```diff
@@ -1,14 +1,48 @@
...
+like [yamux](yamux), mplex uses a message-based framing layer that enables it to multiplex
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:25]: Both mplex and yamux define a framing layer.

File: content/concepts/multiplex/mplex.md [hunk 6]
```diff
@@ -1,14 +1,48 @@
...
+different types of data streams, including stream-oriented data, datagrams, and other types
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:26]: mplex doesn't have datagrams.

File: content/concepts/multiplex/mplex.md [hunk 7]
```diff
@@ -1,14 +1,48 @@
...
+of messages. This flexibility comes at the cost of some performance and reliability, as
+mplex is less efficient at transferring large amounts of data than other multiplexing
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:28]: That's not really true, if you look at raw throughput, mplex is not doing badly. It's able to utilize the full underlying TCP connection.

File: content/concepts/multiplex/mplex.md [hunk 8]
```diff
@@ -1,14 +1,48 @@
...
+protocols that support backpressure.
+
+Additionally, mplex has no limits on how many streams a peer can open. This allows mplex
+to support a wide range of applications and use cases, but it also means that mplex may
+not be as efficient as other multiplexing protocols that limit the number of streams that
+can be opened.
+
+{{< alert icon="" context="">}}
+**Yamux should be used over mplex in libp2p**. It is well-suited for applications that
+require fast, reliable data transfer, such as real-time communication or streaming media,
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:38]: My comment from the yamux document about real-time communication applies here too.

File: content/concepts/multiplex/mplex.md [hunk 9]
```diff
@@ -1,14 +1,48 @@
...
+as it is optimized for low-latency, high-bandwidth environments.
+
+Until recently, the reason mplex was still supported was compatibility with js-libp2p,
+which didn't have yamux support.
+Now that
+[js-libp2p has gained yamux support](https://github.com/ChainSafe/js-libp2p-yamux/releases/tag/v1.0.0),
+you should only support mplex if you need to deal with legacy nodes.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:45]:
```suggestion
mplex should only be used to provide backwards-compatibility with legacy nodes.
```

**Comment by DannyS03** [comment]
@marten-seemann can be checked again.

File: content/concepts/multiplex/mplex.md [hunk 10]
```diff
@@ -1,14 +1,42 @@
 ---
 title: "mplex"
-description: "mplex is a protocol developed for libp2p."
+description: "mplex is a stream multiplexer developed for libp2p."
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:3]:
```suggestion
description: "mplex is a simple stream multiplexer developed for libp2p."
```

File: content/concepts/multiplex/mplex.md [hunk 11]
```diff
@@ -1,14 +1,42 @@
...
-This section is incomplete, and many of the articles are stubs. To help fill in
-the gaps, please see the issues linked in each article to add your input and
-help us prioritize the outstanding work.
+mplex is a stream multiplexer designed for libp2p.
+It is a simple protocol that does not provide many features offered by other
+stream multiplexers. Notably, mplex does not provide backpressure at the protocol
+level.
+
+mplex runs over a reliable, ordered pipe between two peers, such as a TCP connection
+or a Unix pipe. Peers can open, write to, close, and reset a stream by sending messages
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:15]: We don't really use Unix pipes anywhere, so it's probably not worth mentioning.

File: content/concepts/multiplex/mplex.md [hunk 12]
```diff
@@ -1,14 +1,42 @@
...
+with the appropriate [flag value](https://github.com/libp2p/specs/tree/master/mplex#flag-values).
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:16]: It doesn't really matter in a documentation how these things are encoded.

File: content/concepts/multiplex/mplex.md [hunk 13]
```diff
@@ -1,14 +1,42 @@
...
+It uses a message-based framing layer like [yamux](yamux), enabling it to multiplex different
+data streams, including stream-oriented data and other types of messages.
+
+Every message in mplex consists of a header and the prefixed data segment length.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:20]: This is also probably too detailed.

File: content/concepts/multiplex/mplex.md [hunk 14]
```diff
@@ -1,14 +1,42 @@
...
+
+### Drawbacks
+
+mplex does not have any flow control.
+> Backpressure is a mechanism to prevent one peer from overwhelming a slow time consuming the data.
+
+mplex has no limits on how many streams a peer can open. This allows mplex to support a wide
+range of applications and use cases, but it also means that mplex may not be as efficient as other
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:28]: Not really. QUIC has that and I wouldn't say that this limits QUIC's use cases.

File: content/concepts/multiplex/mplex.md [hunk 15]
```diff
@@ -1,14 +1,42 @@
...
+multiplexing protocols that limit the number of streams that can be opened.
+
+{{< alert icon="" context="">}}
+**Yamux should be used over mplex in libp2p**. It is well-suited for applications that
+require fast, reliable data transfer as it is optimized for low-latency, high-bandwidth environments.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:33]:
```suggestion
require fast, reliable data transfer.
```

File: content/concepts/multiplex/mplex.md [hunk 16]
```diff
@@ -1,14 +1,38 @@
...
 
-mplex is a protocol developed for libp2p. The [spec](https://github.com/libp2p/specs/tree/master/mplex) defines a
-simple protocol for multiplexing that is widely supported across libp2p language implementations:
+## What is mplex?
 
-{{< alert icon="💡" context="note">}}
-This section is incomplete, and many of the articles are stubs. To help fill in
-the gaps, please see the issues linked in each article to add your input and
-help us prioritize the outstanding work.
+mplex is a stream multiplexer designed for libp2p.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:9]:
```suggestion
mplex is a simple stream multiplexer that was designed in the early days of libp2p.
```

File: content/concepts/multiplex/mplex.md [hunk 17]
```diff
@@ -1,14 +1,38 @@
...
+It is a simple protocol that does not provide many features offered by other
+stream multiplexers. Notably, mplex does not provide backpressure at the protocol
+level.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:12]:
```suggestion
stream multiplexers. Notably, mplex does not provide flow control, a feature which
is now considered critical for a stream multiplexer.
```

File: content/concepts/multiplex/mplex.md [hunk 18]
```diff
@@ -1,14 +1,38 @@
...
+layer like [yamux](yamux), enabling it to multiplex different
+data streams, including stream-oriented data and other types of messages.
+
+### Drawbacks
+
+mplex does not have any flow control.
+> Backpressure is a mechanism to prevent one peer from overwhelming a slow time consuming the data.
+
+mplex has no limits on how many streams a peer can open. This allows mplex to support a wide
+range of applications and use cases.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:25]:
```suggestion
mplex also doesn't limit how many streams a peer can open.
```

File: content/concepts/multiplex/mplex.md [hunk 19]
```diff
@@ -1,14 +1,38 @@
...
+
+{{< alert icon="" context="">}}
+**Yamux should be used over mplex in libp2p**. It is well-suited for applications that
+require fast, reliable data transfer.
```
Comment (marten-seemann) [inline content/concepts/multiplex/mplex.md:29]:
```suggestion
**Yamux should be used over mplex in libp2p**. As it natively supports flow control, it is better suited for applications that require the transfer of large amounts of data.
```

**Review by marten-seemann — APPROVED** [review]
This should be good to go once the suggestions have been incorporated :)

**Pull request merged**